File: src/lib/crdt/gset.test.ts
import { describe, expect, it } from "vitest";
import GSet from "./gset";

describe("GSet", () => {
    it("should be able to add an element", () => {
        const set = new GSet();

        expect(set.value).toEqual(new Set());
        expect(set.add(1)).toEqual(new Set([1]));
    });

    it("should be able to add multiple elements", () => {
        const set = new GSet();

        expect(set.value).toEqual(new Set());
        expect(set.add(1)).toEqual(new Set([1]));
        expect(set.add(2)).toEqual(new Set([1, 2]));
        expect(set.add(3)).toEqual(new Set([1, 2, 3]));
    });

    it("should be able to add an element multiple times", () => {
        const set = new GSet();

        expect(set.value).toEqual(new Set());
        expect(set.add(1)).toEqual(new Set([1]));
        expect(set.add(1)).toEqual(new Set([1]));
        expect(set.add(1)).toEqual(new Set([1]));
    });

    it("should be able to start with a value", () => {
        const set = new GSet(new Set([1, 2, 3]));

        expect(set.value).toEqual(new Set([1, 2, 3]));
    });

    it("should be able to merge with itself", () => {
        const set = new GSet();

        set.add(1);
        set.add(2);

        expect(set.merge(set)).toEqual(new Set([1, 2]));
    });

    it("should be able to merge with another set", () => {
        const set1 = new GSet();
        const set2 = new GSet();

        set1.add(1);
        set1.add(2);
        set2.add(3);

        expect(set1.merge(set2)).toEqual(new Set([1, 2, 3]));
    });

    it("should be able to be converted to JSON", () => {
        const set = new GSet();

        set.add(1);
        set.add(2);
        set.add(3);

        expect(JSON.stringify(set)).toEqual("[1,2,3]");
    });
});

File: server/src/test/kotlin/pt/up/fe/sdle/crdt/GSetTest.kt
package pt.up.fe.sdle.crdt

import kotlin.test.Test
import kotlin.test.assertEquals

class GSetTest {
    @Test
    fun `GSet - add an element`() {
        val set = GSet<Int>()

        assertEquals(setOf(), set.value)
        assertEquals(setOf(1), set.add(1))
    }

    @Test
    fun `GSet - add multiple elements`() {
        val set = GSet<Int>()

        assertEquals(setOf(), set.value)
        assertEquals(setOf(1), set.add(1))
        assertEquals(setOf(1, 2), set.add(2))
        assertEquals(setOf(1, 2, 3), set.add(3))
    }

    @Test
    fun `GSet - add an element multiple times`() {
        val set = GSet<Int>()

        assertEquals(setOf(), set.value)
        assertEquals(setOf(1), set.add(1))
        assertEquals(setOf(1), set.add(1))
        assertEquals(setOf(1), set.add(1))
    }

    @Test
    fun `GSet - merge with itself`() {
        val set = GSet<Int>()

        set.add(1)
        set.add(2)

        assertEquals(setOf(1, 2), set.merge(set))
    }

    @Test
    fun `GSet - merge with another set`() {
        val set1 = GSet<Int>()
        val set2 = GSet<Int>()

        set1.add(1)
        set1.add(2)
        set2.add(3)

        assertEquals(setOf(1, 2, 3), set1.merge(set2))
    }
}
